### ISSUE TITLE
Unknown type from Lazy.lock()

### ISSUE THREAD

**Issue by Timmmm** [rust-analyzer/rust-analyzer#2880]
I'm using [`once_cell`](https://crates.io/crates/once_cell)'s [`Lazy`](https://docs.rs/once_cell/1.3.1/once_cell/sync/struct.Lazy.html) type, like this:

```
use once_cell::sync::Lazy;

struct State {
}

static STATE: Lazy<Arc<Mutex<State>>> = Lazy::new(|| {
  Arc::new(Mutex::new(Default::default()))
});

fn foo() {
  let a = STATE.lock();
```

If I hover `STATE` then it does show the type as `static STATE: Lazy<Arc<Mutex<State>>>`, however completion does not work (it doesn't offer `.lock()`), and if you hover `a` or `lock` it says `{unknown}`. I guess this is something to do with the `Deref` trait.

PS: This is with `master` from today (de2409702).

**Comment by flodiebold** [comment]
The `Deref` implementation is
```rust
impl<T, F: FnOnce() -> T> Deref for Lazy<T, F>
```
and `Lazy` is
```rust
pub struct Lazy<T, F = fn() -> T>
```
so the problem here is that we don't provide the built-in implementation of `FnOnce` for function pointers yet.

**Comment by Timmmm** [comment]
Is this hard to implement? I might have a go if you can point me in the right direction.

**Comment by flodiebold** [comment]
It's not trivial, but it's very similar to the existing built-in implementation of the `Fn` traits for closures, so it should be doable :slightly_smiling_face: 

There's two separate pieces of data that Chalk asks for for these impls: the impl itself, and the values of associated types (the `Output` type in this case). These get identified by the [`Impl` type](https://github.com/rust-analyzer/rust-analyzer/blob/a878f39b5eea8876ee6238eca9b935cc32656862/crates/ra_hir_ty/src/traits.rs#L341-L346) and the [`AssocTyValue` type](https://github.com/rust-analyzer/rust-analyzer/blob/a878f39b5eea8876ee6238eca9b935cc32656862/crates/ra_hir_ty/src/traits.rs#L356-L361). So the first step would be to add a new variant for each of these. The data we need here are the `FnTrait` (since we want to implement all of them) and the number of arguments (since we need a separate impl for each possible number of arguments to the fn pointer).

The second (and biggest) step is implementing [`impl_datum` and `associated_ty_value`](https://github.com/rust-analyzer/rust-analyzer/blob/a878f39b5eea8876ee6238eca9b935cc32656862/crates/ra_hir_ty/src/traits/builtin.rs#L48-L66) for these. These build the data describing the impl for Chalk. You'll need functions very similar to `closure_fn_trait_impl_datum` and `closure_fn_trait_output_assoc_ty_value`: The closure impls look something like this:
```rust
impl<T, U, V> Fn<(T, U)> for closure<fn(T, U) -> V> { Output = V }
```
and we need
```rust
impl<T, U, V> Fn<(T, U)> for fn(T, U) -> V { Output = V }
```
So mostly the difference is where we get `num_args` from (directly from the enum variant added in step 1), and that we don't wrap `sig_ty` in the closure type.

The third step is actually returning these impls when Chalk asks for them. This happens in [`get_builtin_impls`](https://github.com/rust-analyzer/rust-analyzer/blob/a878f39b5eea8876ee6238eca9b935cc32656862/crates/ra_hir_ty/src/traits/builtin.rs#L24-L46). We need to check the given type to see if it's a function pointer type, check the prerequisites the same way we do for the closure impl, and then construct the right `Impl` variant and pass it to the callback.
This `check_closure_fn_trait_impl_prerequisites` thing happens because once we've told Chalk there is an impl, we can't later say it doesn't exist after all. So we have to make sure that we can actually construct it, which means that the respective `Fn` trait actually has `FnOnce` as a super trait, and `FnOnce` has an `Output` associated type. (Of course in reality this should always be the case, but since these are defined in the standard library we can't technically rely on it.)

The last thing would just be adding a test a bit similar to [here](https://github.com/rust-analyzer/rust-analyzer/blob/a878f39b5eea8876ee6238eca9b935cc32656862/crates/ra_hir_ty/src/tests/traits.rs#L1283); you could try calling `call_once` on a function pointer in the test.

**Comment by lnicola** [comment]
Fixed by #5124.
